File: mobile/kmp/ui/src/commonMain/kotlin/com/atls/hyperion/ui/theme/typography/Styles.kt
package com.atls.hyperion.ui.theme.typography

import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily

class TextStyles(fontFamily: FontFamily) {
    val display = Display(
        lg = TextStyle(
            fontFamily = fontFamily,
            fontSize = FontSize.xl6,
            lineHeight = LineHeight.xl4,
            fontWeight = FontWeight.semiBold,
            letterSpacing = LetterSpacing.md
        ),
        md = TextStyle(
            fontFamily = fontFamily,
            fontSize = FontSize.xl5,
            lineHeight = LineHeight.xl3,
            fontWeight = FontWeight.semiBold,
            letterSpacing = LetterSpacing.md
        ),
        sm = TextStyle(
            fontFamily = fontFamily,
            fontSize = FontSize.xl4,
            lineHeight = LineHeight.xl2,
            fontWeight = FontWeight.semiBold,
            letterSpacing = LetterSpacing.md
        )
    )
    val heading = Heading(
        lg = TextStyle(
            fontFamily = fontFamily,
            fontSize = FontSize.xl3,
            lineHeight = LineHeight.xl,
            fontWeight = FontWeight.medium,
            letterSpacing = LetterSpacing.md
        ),
        md = TextStyle(
            fontFamily = fontFamily,
            fontSize = FontSize.xl2,
            lineHeight = LineHeight.lg,
            fontWeight = FontWeight.medium,
            letterSpacing = LetterSpacing.md
        ),
        sm = TextStyle(
            fontFamily = fontFamily,
            fontSize = FontSize.xl,
            lineHeight = LineHeight.md,
            fontWeight = FontWeight.medium,
            letterSpacing = LetterSpacing.md
        )
    )
    val title = Title(
        lg = TextStyle(
            fontFamily = fontFamily,
            fontSize = FontSize.lg,
            lineHeight = LineHeight.sm,
            fontWeight = FontWeight.medium,
            letterSpacing = LetterSpacing.md
        ),
        md = TextStyle(
            fontFamily = fontFamily,
            fontSize = FontSize.md,
            lineHeight = LineHeight.sm,
            fontWeight = FontWeight.medium,
            letterSpacing = LetterSpacing.md
        )
    )
    val body = Body(
        lg = TextStyle(
            fontFamily = fontFamily,
            fontSize = FontSize.md,
            lineHeight = LineHeight.sm,
            fontWeight = FontWeight.regular,
            letterSpacing = LetterSpacing.md
        ),
        md = TextStyle(
            fontFamily = fontFamily,
            fontSize = FontSize.sm,
            lineHeight = LineHeight.xs,
            fontWeight = FontWeight.regular,
            letterSpacing = LetterSpacing.md
        ),
        sm = TextStyle(
            fontFamily = fontFamily,
            fontSize = FontSize.xs,
            lineHeight = LineHeight.xs2,
            fontWeight = FontWeight.regular,
            letterSpacing = LetterSpacing.md
        )
    )
    val label = Label(
        lg = TextStyle(
            fontFamily = fontFamily,
            fontSize = FontSize.md,
            lineHeight = LineHeight.xs,
            fontWeight = FontWeight.medium,
            letterSpacing = LetterSpacing.lg
        ),
        md = TextStyle(
            fontFamily = fontFamily,
            fontSize = FontSize.sm,
            lineHeight = LineHeight.xs,
            fontWeight = FontWeight.medium,
            letterSpacing = LetterSpacing.lg
        ),
        sm = TextStyle(
            fontFamily = fontFamily,
            fontSize = FontSize.xs,
            lineHeight = LineHeight.xs2,
            fontWeight = FontWeight.medium,
            letterSpacing = LetterSpacing.lg
        )
    )
    val caption: TextStyle = TextStyle(
        fontFamily = fontFamily,
        fontSize = FontSize.xs2,
        lineHeight = LineHeight.xs2,
        fontWeight = FontWeight.regular,
        letterSpacing = LetterSpacing.xl
    )
    val overline: TextStyle = TextStyle(
        fontFamily = fontFamily,
        fontSize = FontSize.xs3,
        lineHeight = LineHeight.xs3,
        fontWeight = FontWeight.medium,
        letterSpacing = LetterSpacing.xl
    )
}

File: ui/theme/src/tokens/typography/styles.ts
import type { FontSizeValue }      from './interfaces.js'
import type { FontWeightValue }    from './interfaces.js'
import type { LetterSpacingValue } from './interfaces.js'
import type { LineHeightValue }    from './interfaces.js'
import type { TypographyStyle }    from './interfaces.js'

import { scienceGothicFontFamily } from './font-faces.js'
import { fontSizes }               from './font-sizes.js'
import { fontWeights }             from './font-weights.js'
import { letterSpacings }          from './letter-spacings.js'
import { lineHeights }             from './line-heights.js'

const createTextStyle = (
  fontSize: FontSizeValue,
  lineHeight: LineHeightValue,
  fontWeight: FontWeightValue,
  letterSpacing: LetterSpacingValue
): TypographyStyle => ({
  fontFamily: scienceGothicFontFamily,
  fontSize,
  lineHeight,
  fontWeight,
  letterSpacing,
})

export const textStyles = {
  display: {
    lg: createTextStyle(fontSizes.xl6, lineHeights.xl4, fontWeights.semiBold, letterSpacings.md),
    md: createTextStyle(fontSizes.xl5, lineHeights.xl3, fontWeights.semiBold, letterSpacings.md),
    sm: createTextStyle(fontSizes.xl4, lineHeights.xl2, fontWeights.semiBold, letterSpacings.md),
  },
  heading: {
    lg: createTextStyle(fontSizes.xl3, lineHeights.xl, fontWeights.medium, letterSpacings.md),
    md: createTextStyle(fontSizes.xl2, lineHeights.lg, fontWeights.medium, letterSpacings.md),
    sm: createTextStyle(fontSizes.xl, lineHeights.md, fontWeights.medium, letterSpacings.md),
  },
  title: {
    lg: createTextStyle(fontSizes.lg, lineHeights.sm, fontWeights.medium, letterSpacings.md),
    md: createTextStyle(fontSizes.md, lineHeights.sm, fontWeights.medium, letterSpacings.md),
  },
  body: {
    lg: createTextStyle(fontSizes.md, lineHeights.sm, fontWeights.regular, letterSpacings.md),
    md: createTextStyle(fontSizes.sm, lineHeights.xs, fontWeights.regular, letterSpacings.md),
    sm: createTextStyle(fontSizes.xs, lineHeights.xs2, fontWeights.regular, letterSpacings.md),
  },
  label: {
    lg: createTextStyle(fontSizes.md, lineHeights.xs, fontWeights.medium, letterSpacings.lg),
    md: createTextStyle(fontSizes.sm, lineHeights.xs, fontWeights.medium, letterSpacings.lg),
    sm: createTextStyle(fontSizes.xs, lineHeights.xs2, fontWeights.medium, letterSpacings.lg),
  },
  caption: createTextStyle(fontSizes.xs2, lineHeights.xs2, fontWeights.regular, letterSpacings.xl),
  overline: createTextStyle(fontSizes.xs3, lineHeights.xs3, fontWeights.medium, letterSpacings.xl),
} as const
